### PR TITLE
EOS-16167 Refactor wait_for_all.

### PR DESCRIPTION
**Pull request by gshipra**
EOS-16167 Refactor wait_for_all and wait_for_any.

Problem:
wait_for_all() and wait_for_any() in DeliveryHerald
class, both functions have replicated code.
Solution:
Created a new function _verify_delivered() in
DeliveryHerald class. This function has common logic 
for checking if delievery has happened. wait_for_any()
and wait_for_all() functions will replace common code
with _verify_delivered() function.
Test steps:
make ran the test_delivery_herald.py
and all tests passed.
Bootstrap passed.

Signed-off-by: Shipra Gupta <shipra.gupta@seagate.com>

### CODE REVIEW

**Comment by gshipra** [comment]
@mssawant @chumakd - Please provide your feedback for the changes done. Let me know what else I can do for testing this fix. There are some tests which were automatically at make time.

**Comment by mssawant** [comment]
@gshipra, kindly reformat the commit message as,
Problem: <short summary of the problem>
<Additional details of the problem>
Solution: 
<Solution for the problem>

File: hax/hax/handler.py [hunk 1]
```diff
@@ -110,7 +110,8 @@ def pull_msg():
                                      status=ServiceHealth.FAILED)])
                         LOG.debug('waiting for broadcast of %s for ep: %s',
                                   ids, item.remote_rpc_endpoint)
-                        self.herald.wait_for_all(HaLinkMessagePromise(ids))
+                        self.herald.wait_for_delivery(
```
Comment (mssawant) [inline hax/hax/handler.py:113]: I would still prefer to have 2 separate explicitly named functions, wait_for_any() and wait_for_all(). You can rather take out the common content of the functions into a separate local function which can be invoked from both of them.

File: hax/hax/motr/delivery.py [hunk 2]
```diff
@@ -101,6 +79,8 @@ def wait_for_all(self,
                         confirmed_msgs)
                 del self.waiting_clients[promise]
                 promise.exclude_ids(confirmed_msgs)
+                if option.lower() == "any":
```
Comment (mssawant) [inline hax/hax/motr/delivery.py:82]: Let's avoid this and keep the 2 explicitly named functions but common out the content as mentioned above.

**Comment by gshipra** [comment]
> @gshipra, kindly reformat the commit message as,
> Problem: `<short summary of the problem>`
> `<Additional details of the problem>`
> Solution:
> `<Solution for the problem>`

Reformatted the message.

File: hax/hax/handler.py (hunk 1 above)
Comment (gshipra) [reply hax/hax/handler.py]: Done.

File: hax/hax/motr/delivery.py [hunk 3]
```diff
@@ -46,6 +46,24 @@ def __init__(self):
         self.waiting_clients: Dict[HaLinkMessagePromise, Condition] = {}
         self.lock = Lock()
 
+    def check_delivery(self,
```
Comment (mssawant) [inline hax/hax/motr/delivery.py:49]: [suggestion]: s/check_delivery/is_delivered/

File: hax/hax/motr/delivery.py [hunk 4]
```diff
@@ -46,6 +46,24 @@ def __init__(self):
...
+                       promise: HaLinkMessagePromise,
+                       timeout_sec: float):
+        """
+        Check if any message in promise._ids are reported by Motr
+        as delivered. Calling function should hold the lock.
```
Comment (mssawant) [inline hax/hax/motr/delivery.py:54]: [suggestion]: It would be good to mention the lock here (self.lock) in documentation.

**Review by mssawant — COMMENTED** [review]
Minor comments, looks good.
@gshipra, what kind of testing is done on the patch, is the boot strapping working fine?

**Comment by gshipra** [comment]
> Minor comments, looks good.
> @gshipra, what kind of testing is done on the patch, is the boot strapping working fine?

Bootstrap is working

 hctl bootstrap --mkfs ~/CDF.yaml
2021-01-07 19:18:55: Generating cluster configuration... OK
2021-01-07 19:18:57: Starting Consul server on this node.......... OK
2021-01-07 19:19:04: Importing configuration into the KV store... OK
2021-01-07 19:19:04: Starting Consul on other nodes... OK
2021-01-07 19:19:05: Updating Consul configuraton from the KV store... OK
2021-01-07 19:19:05: Waiting for the RC Leader to get elected....... OK
2021-01-07 19:19:10: Starting Motr (phase1, mkfs)... OK
2021-01-07 19:19:18: Starting Motr (phase1, m0d)... OK
2021-01-07 19:19:30: Starting Motr (phase2, mkfs)... OK
2021-01-07 19:19:37: Starting Motr (phase2, m0d)... OK
2021-01-07 19:19:49: Checking health of services... OK


hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x2f 'the pool'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x4f 'default': 'the pool' None None
Services:
    localhost  (RC)
    [started]  hax        0x7200000000000001:0x6   10.230.245.192@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x9   10.230.245.192@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0xc   10.230.245.192@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x29  10.230.245.192@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x2c  10.230.245.192@tcp:12345:4:2

File: hax/hax/motr/delivery.py (hunk 4 above)
Comment (gshipra) [reply hax/hax/motr/delivery.py]: updated lock to self.lock

**Review by mssawant — APPROVED** [review]

**Comment by chumakd** [comment]
Even though the blocks of code in `wait_for_all()` and `wait_for_any()` look similar, they're aren't exactly identical. I suspect the differences aren't just stylistic, but they implement slightly different business logic. I'd suggest waiting for @knekrasov 's opinion as well before merging it.

**Review by chumakd — CHANGES_REQUESTED** [review]
Even though the blocks of code in `wait_for_all()` and `wait_for_any()` look similar, they're aren't exactly identical. I suspect the differences aren't just stylistic, but they implement slightly different business logic. I'd suggest waiting for @knekrasov 's opinion as well before merging it.

File: hax/hax/motr/delivery.py (hunk 3 above)
Comment (knekrasov) [reply hax/hax/motr/delivery.py:49]: My strong opinion is that the method name should start with `_` prefix (to mark that it is a private one, no one is expected to invoke it outside of this class; as a bonus no docstring is required for a private method 😁).

`_is_delivered()` looks better than `check_delivery` but it doesn't return a `bool` value (although `is` is usually a sign of boolean). So I would recommend something like `_ensure_delivered()`.

In fact the method does quite a lot of different things, so it is hard to name it properly. Let's just choose a name that would mislead the developers in a minimal way.

**Review by knekrasov — CHANGES_REQUESTED** [review]
Looks good to me. Let's just improve the naming.

**Comment by gshipra** [comment]
> Looks good to me. Let's just improve the naming.

Changed the function name to _verify_delivered

**Comment by knekrasov** [comment]
@gshipra Last request from my side 😄 
Can you please rebase your branch against main branch (to avoid merge commit) and then squash your changes into a single commit? I don't think it makes sense to store more commits given the change is quite atomic and self-contained.

**Comment by gshipra** [comment]
> @gshipra Last request from my side 😄
> Can you please rebase your branch against main branch (to avoid merge commit) and then squash your changes into a single commit? I don't think it makes sense to store more commits given the change is quite atomic and self-contained.

Could you please tell me the steps for rebase (to avoid merge commit)

**Comment by knekrasov** [comment]
@gshipra 
Roughly and very high level:

## Rebase against the fresh main
1. `git fetch origin` (assuming that `origin` is the name of the default remote that points to `https://github.com/Seagate/cortx-hare.git`
2. `git checkout EOS-16167`
3. `git rebase origin/main`

## Squash commits
1. `git checkout EOS-16167`
2. `git rebase --interactive HEAD~3`
More info on squashing commits via interactive rebase can be found here: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

**Comment by gshipra** [comment]
> @gshipra
> Roughly and very high level:
> 
> ## Rebase against the fresh main
> 1. `git fetch origin` (assuming that `origin` is the name of the default remote that points to `https://github.com/Seagate/cortx-hare.git`
> 2. `git checkout EOS-16167`
> 3. `git rebase origin/main`
> 4. `git push -f` (rebase means rewriting the commit history; that's why force flag is necessary)
> 
> ## Squash commits
> 1. `git checkout EOS-16167`
> 2. `git rebase --interactive HEAD~3`
> 3. `git push -f` (rebase means rewriting the commit history; that's why force flag is necessary)
> 
> More info on squashing commits via interactive rebase can be found here: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

Done rebasing and squashing.

**Review by knekrasov — APPROVED** [review]
Nice work! Thank you

**Pull request merged**